prow/plugins/mergecommitblocker/mergecommitblocker.go

```diff
@@ -0,0 +1,127 @@
+	fixmeBody   = "Your pull request contains merge commits, which are not allowed in this repository. Use `git rebase` to reapply your commits on top of the target branch."
+)
+
+// init registers out plugin as a pull request handler
+func init() {
+	plugins.RegisterPullRequestHandler(PluginName, handlePullRequest, helpProvider)
+}
+
+// helpProvider provides information on the plugin
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf("The merge commit blocker plugin adds the %s label to pull requests that contain merge commits", labels.MergeCommits),
+	}, nil
+}
+
+type githubClient interface {
+	AddLabel(org, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	CreateComment(org, repo string, number int, comment string) error
+}
+
+type pruneClient interface {
+	PruneComments(func(ic github.IssueComment) bool)
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened &&
+		pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	cp, err := pc.CommentPruner()
+	if err != nil {
+		return err
+	}
+	return handle(pc.GitHubClient, pc.GitClient, cp, pc.Logger, &pre)
+}
+
+func handle(ghc githubClient, gc *git.Client, cp pruneClient, log *logrus.Entry, pre *github.PullRequestEvent) error {
+	var (
+		org  = pre.PullRequest.Base.Repo.Owner.Login
+		repo = pre.PullRequest.Base.Repo.Name
+		num  = pre.PullRequest.Number
+	)
+
+	// Clone the repo, checkout the PR.
+	r, err := gc.Clone(fmt.Sprintf("%s/%s", org, repo))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(num); err != nil {
+		return err
+	}
+	// We are guaranteed to have both Base.SHA and Head.SHA
+	target, head := pre.PullRequest.Base.SHA, pre.PullRequest.Head.SHA
+	existMergeCommits, err := r.MergeCommitsExistBetween(target, head)
+	if err != nil {
+		return err
+	}
+	issueLabels, err := ghc.GetIssueLabels(org, repo, num)
+	if err != nil {
+		return err
+	}
+	hasLabel := github.HasLabel(labels.MergeCommits, issueLabels)
+	if hasLabel && !existMergeCommits {
+		log.Infof("Removing %q Label for %s/%s#%d", labels.MergeCommits, org, repo, num)
+		if err := ghc.RemoveLabel(org, repo, num, labels.MergeCommits); err != nil {
+			return err
+		}
+		msg := plugins.FormatResponse(pre.PullRequest.User.Login, blockedBody, fixmeBody)
+		return ghc.CreateComment(org, repo, pre.Number, msg)
```

Comment:
I'm pretty sure this is backwards. We should post a comment when we add the label, not when we remove it (and similarly prune comments when we remove the label instead of when we add it).
Please validate this behavior with unit tests.

Reply:
My bad on copying/pasting...